// what-core/dist/jsx-dev-runtime.js
// packages/core/src/h.js
var EMPTY_OBJ = /* @__PURE__ */ Object.create(null);
function h(tag, props, ...children) {
  props = props || EMPTY_OBJ;
  const flat = flattenChildren(children);
  const key = props.key ?? null;
  if (props.key !== void 0) {
    props = { ...props };
    delete props.key;
  }
  return { tag, props, children: flat, key, _vnode: true };
}
function Fragment({ children }) {
  return children;
}
function flattenChildren(children) {
  const out = [];
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    if (child == null || child === false || child === true) continue;
    if (Array.isArray(child)) {
      out.push(...flattenChildren(child));
    } else if (typeof child === "object" && child._vnode) {
      out.push(child);
    } else if (typeof child === "function") {
      out.push(child);
    } else {
      out.push(String(child));
    }
  }
  return out;
}

// packages/core/src/jsx-dev-runtime.js
function jsxDEV(type, props, key) {
  if (props == null) return h(type, null);
  const { children, ...rest } = props;
  if (key !== void 0) rest.key = key;
  if (children === void 0) return h(type, rest);
  if (Array.isArray(children)) return h(type, rest, ...children);
  return h(type, rest, children);
}
var jsx = jsxDEV;
var jsxs = jsxDEV;
export {
  Fragment,
  jsx,
  jsxDEV,
  jsxs
};
